### PR TITLE
Revert state

### PR DESCRIPTION
**Pull request by AvishaiW**
this PR adds:

-  IPC call: remove contract- removes all the contents of a given contract in the local DB.
- IPC call: remove_deltas- remove all deltas with the indices given by the call.
- remove_delta ocall which removes the delta from inside of the enclave.
- DB call: delete_contract which deletes all of the contents of a contract in one call.
- tests which check all of the above

### CODE REVIEW

File: enigma-core/app/src/db/dal.rs [hunk 1]
```diff
@@ -322,6 +351,29 @@ mod test {
...
     }
 
+    #[test]
+    fn test_delete_contract_contents() {
+        let (mut db, _dir) = create_test_db();
+
+        let arr = [5u8; 32];
+        let v = b"Enigma";
+        db.create(&Array32u8(arr), &v[..]).unwrap();
+        db.delete_contract(&Array32u8(arr)).unwrap();
```
Comment (moriaab) [inline enigma-core/app/src/db/dal.rs:361]: Please add a verification that the contract was indeed deleted.

File: enigma-core/app/src/db/dal.rs [hunk 2]
```diff
@@ -322,6 +351,29 @@ mod test {
...
+        let (mut db, _dir) = create_test_db();
+        let addr: ContractAddress = [2u8; 32].into();
+        let dk_code = DeltaKey::new(addr, Stype::ByteCode);
+        let dk_delta = DeltaKey::new(addr, Stype::Delta(0));
+        let v_code = b"This is a Contract";
+        let v_delta = b"This is a Delta";
+        db.create(&dk_code, &v_code[..]).unwrap();
+        db.create(&dk_delta, &v_delta[..]).unwrap();
+        db.delete_contract(&dk_code).unwrap();
+    }
```
Comment (moriaab) [inline enigma-core/app/src/db/dal.rs:389]: Please add a verification that the contract was indeed deleted.

File: enigma-core/enclave/src/lib.rs [hunk 3]
```diff
@@ -279,9 +279,14 @@ unsafe fn store_delta_and_state(db_ptr: *const RawPointer, delta: &Option<Encryp
         Some(d) => {
             let enc_state = km_t::encrypt_state(state.clone())?;
             enigma_runtime_t::ocalls_t::save_delta(db_ptr, d)?;
-            // todo: in case the delta was stored and the state wasn't, we should revert the delta- create a new ocall
-            enigma_runtime_t::ocalls_t::save_state(db_ptr, &enc_state)?;
-            Ok(())
+            match enigma_runtime_t::ocalls_t::save_state(db_ptr, &enc_state) {
```
Comment (moriaab) [inline enigma-core/enclave/src/lib.rs]: I think that it is better to do: `if let Err(e) = enigma_runtime_t::ocalls_t::save_state(db_ptr, &enc_state) {...}` or use `is_err` than `match` here.

Comment (AvishaiW) [reply enigma-core/enclave/src/lib.rs]: https://github.com/enigmampc/enigma-core/pull/224/commits/cfee01897053a154c5ba21683776d254758143f0

File: enigma-core/app/src/esgx/ocalls_u.rs [hunk 4]
```diff
@@ -197,6 +198,32 @@ pub unsafe extern "C" fn ocall_get_deltas(db_ptr: *const RawPointer, addr: &Cont
     }
 }
 
+#[no_mangle]
+pub unsafe extern "C" fn ocall_remove_delta(db_ptr: *const RawPointer,
+                                            contract_address: &ContractAddress, _delta_index: *const u32) -> EnclaveReturn {
```
Comment (moriaab) [inline enigma-core/app/src/esgx/ocalls_u.rs]: Very minor, but I don't understand why you use a preceding underscore in name `_delta_index`. Use `delta_index_` if you want to differ this name from `delta_index`. The convention for preceding underscores is unused values.

File: enigma-core/app/src/esgx/ocalls_u.rs [hunk 5]
```diff
@@ -197,6 +198,32 @@ pub unsafe extern "C" fn ocall_get_deltas(db_ptr: *const RawPointer, addr: &Cont
...
+            return EnclaveReturn::OcallDBError
+        }
+    };
+    match db.delete(&key) {
+        Ok(_) => EnclaveReturn::Success,
+        Err(e) => {
+            match e.downcast::<DBErr>() {
+                Ok(_) =>  EnclaveReturn::Success,
+                Err(_) => {
+                    println!("Failed removing delta: {:?}", &key);
```
Comment (moriaab) [inline enigma-core/app/src/esgx/ocalls_u.rs]: I think that `error!` should replace`println!` here. Isn't it worth to print `e` as part of the message?

File: enigma-core/app/src/esgx/ocalls_u.rs (hunk 4 above)
Comment (AvishaiW) [reply enigma-core/app/src/esgx/ocalls_u.rs]: fixed this. (in `ocall_new_delta` as well)

File: enigma-core/app/src/esgx/ocalls_u.rs (hunk 5 above)
Comment (AvishaiW) [reply enigma-core/app/src/esgx/ocalls_u.rs]: fixed this. (in `ocall_new_delta` as well)

File: enigma-core/app/src/networking/ipc_listener.rs [hunk 6]
```diff
@@ -246,16 +250,34 @@ pub(self) mod handling {
...
         // since a new delta and bytecode were added, the state is no longer updated
         db.update_state_status(false);
         let result = IpcResults::Status(status);
         Ok(IpcResponse::UpdateNewContractOnDeployment { address, result })
     }
 
+    #[logfn(INFO)]
+    pub fn remove_contract(db: &mut DB, address: String) -> ResponseResult {
+        let addr_arr = ContractAddress::from_hex(&address)?;
+        // the key_type is irrelevant
```
Comment (moriaab) [inline enigma-core/app/src/networking/ipc_listener.rs]: This comment is not clear

File: enigma-core/app/src/networking/ipc_listener.rs [hunk 7]
```diff
@@ -270,25 +292,64 @@ pub(self) mod handling {
...
-                overall_status = FAILED;
-                FAILED
+                overall_status = Status::Failed;
+                Status::Failed
             } else {
-                PASSED
+                Status::Passed
             };
-            let key = Some(deltakey.key_type.unwrap_delta());
+            let key = Some(deltakey.key_type.unwrap_delta() as i32);
```
Comment (moriaab) [inline enigma-core/app/src/networking/ipc_listener.rs]: This conversion is not safe. I understand that the motivation is to be able to return a special sign `-1`, but maybe I would go for i64.

File: enigma-core/app/src/networking/messages.rs [hunk 8]
```diff
@@ -177,7 +186,7 @@ pub struct IpcTask {
 pub struct IpcStatusResult {
     pub address: String,
     #[serde(skip_serializing_if = "Option::is_none")]
-    pub key: Option<u32>,
+    pub key: Option<i32>,
```
Comment (moriaab) [inline enigma-core/app/src/networking/messages.rs]: Please see my comment regarding this type

File: enigma-runtime-t/src/ocalls_t.rs [hunk 9]
```diff
@@ -163,6 +185,14 @@ pub mod tests {
         assert_eq!(enc, ret);
     }
 
+    pub fn test_remove_delta(db_ptr: *const RawPointer) {
+        let contract_address = b"test_delta_removal".sha256();
+        let (start, end) = (0, 1);
+        let deltas = save_deltas(db_ptr, start, end, &contract_address);
+        let res = remove_delta(db_ptr, deltas.last().unwrap());
+        assert_eq!(res.unwrap(), ());
```
Comment (moriaab) [inline enigma-runtime-t/src/ocalls_t.rs]: Very minor. You may use `assert(res.is_ok())`

**Review by moriaab — APPROVED** [review]
Throughout this PR there is this code 
`                match e.downcast::<DBErr>() {
                    Ok(_) =>  Ok(IpcResults::Status(Status::Passed)),
                    Err(_) => Ok(IpcResults::Status(Status::Failed)),
                }
`
This is not trivial why DBErr is a correct situation, therefore I suggest to comment on this. In addition, I suggest to extract this code to a different function.

File: enigma-core/app/src/esgx/ocalls_u.rs (hunk 4 above)
Comment (AvishaiW) [reply enigma-core/app/src/esgx/ocalls_u.rs]: elaborated

File: enigma-core/app/src/networking/ipc_listener.rs (hunk 6 above)
Comment (AvishaiW) [reply enigma-core/app/src/networking/ipc_listener.rs]: elaborated

File: enigma-core/app/src/networking/ipc_listener.rs (hunk 7 above)
Comment (AvishaiW) [reply enigma-core/app/src/networking/ipc_listener.rs]: fixed

File: enigma-core/app/src/networking/messages.rs (hunk 8 above)
Comment (AvishaiW) [reply enigma-core/app/src/networking/messages.rs]: fixed

File: enigma-runtime-t/src/ocalls_t.rs (hunk 9 above)
Comment (AvishaiW) [reply enigma-runtime-t/src/ocalls_t.rs]: changed. Thanks

**Pull request merged**